File: src/programmers/level2/스킬_트리.java
package programmers.level2;

public class 스킬_트리 {
    public static void main(String[] args) {
        String skill = "CBD";
        String[] skill_trees = {"BACDE", "CBADF", "AECB", "BDA"};
        solution(skill, skill_trees);
    }

    public static int solution(String skill, String[] skill_trees) {
        char[] sk = skill.toCharArray();
        int answer = 0;

        for (String tree : skill_trees) {
            String str = "";
            for (int i = 0; i < tree.length(); i++) {
                for (char ch : sk) {
                    if (ch == tree.charAt(i)) {
                        str += Character.toString(tree.charAt(i));
                    }
                }
            }

            if (str.isEmpty())
                answer++;

            if (skill.contains(str)) {
                if (!str.isEmpty()) {
                    if (sk[0] == str.charAt(0))
                        answer++;
                }
            }
        }

        return answer;

        /* 다른 사람 풀이
        int answer = 0;
        ArrayList<String> skillTrees = new ArrayList<String>(Arrays.asList(skill_trees));
        //ArrayList<String> skillTrees = new ArrayList<String>();
        Iterator<String> it = skillTrees.iterator();

        while (it.hasNext()) {
            if (skill.indexOf(it.next().replaceAll("[^" + skill + "]", "")) != 0) {
                it.remove();
            }
        }

        answer = skillTrees.size();

        return answer;
        */
    }
}

File: 프로그래머스/2/49993. 스킬트리/스킬트리.js
function solution(skill, skill_trees) {
    var answer = 0;
    // skill에 포함되지 않은 모든 문자 제거
    var regex = new RegExp(`[^${skill}]`, 'g');

    return skill_trees
        .map((x) => x.replace(regex, ''))
        .filter((x) => {
            return skill.indexOf(x) === 0 || x === "";
        // skill.indexOf(x) === 0는 x가 skill의 시작 부분과 일치하는지를 확인
        // x === ""는 skill에서 필요한 기술이 하나도 없다는 것을 의미
        })
        .length;
}
